src/hypercall.rs

```diff
@@ -77,21 +79,89 @@ pub unsafe fn address_to_hypercall(
 }
 
 /// unlink deletes a name from the filesystem. This is used to handle `unlink` syscalls from the guest.
-/// TODO: UNSAFE AS *%@#. It has to be checked that the VM is allowed to unlink that file!
-pub fn unlink(mem: &MmapMemory, sysunlink: &mut UnlinkParams) {
-	unsafe {
-		sysunlink.ret = libc::unlink(mem.host_address(sysunlink.name).unwrap() as *const i8);
+pub fn unlink(mem: &MmapMemory, sysunlink: &mut UnlinkParams, file_map: &mut UhyveFileMap) {
+	// For more information on how this works, check out the open call implementation.
+	let requested_path = mem.host_address(sysunlink.name).unwrap() as *const i8;
+	let guest_path = unsafe { CStr::from_ptr(requested_path) }.to_str();
+	if let Ok(guest_path) = guest_path {
+		let host_path_option = file_map.get_host_path(guest_path);
+		if let Some(host_path) = host_path_option {
+			let host_path_c_string = CString::new(host_path.as_bytes()).unwrap();
+			let new_host_path = host_path_c_string.as_c_str().as_ptr();
+
+			unsafe {
+				sysunlink.ret = libc::unlink(new_host_path);
+			}
+		} else {
+			error!("The kernel requested to unlink() an unknown path. Rejecting...");
+			sysunlink.ret = -1;
+		}
+	} else {
+		error!("The kernel requested to open() a path that is not valid UTF-8. Rejecting...");
+		sysunlink.ret = -1;
 	}
 }
 
 /// Handles an open syscall by opening a file on the host.
-pub fn open(mem: &MmapMemory, sysopen: &mut OpenParams) {
-	unsafe {
-		sysopen.ret = libc::open(
-			mem.host_address(sysopen.name).unwrap() as *const i8,
-			sysopen.flags,
-			sysopen.mode,
-		);
+pub fn open(
+	mem: &MmapMemory,
+	sysopen: &mut OpenParams,
+	file_map: &mut UhyveFileMap,
+	temp_dir: &TempDir,
+) {
+	// TODO: Keep track of file descriptors internally, just in case the kernel doesn't close them.
+	let requested_path = mem.host_address(sysopen.name).unwrap() as *const i8;
+	let guest_path = unsafe { CStr::from_ptr(requested_path) }.to_str();
+	let mut flags = sysopen.flags & ALLOWED_OPEN_FLAGS;
+
+	if let Ok(guest_path) = guest_path {
+		// Rust deals in UTF-8. C doesn't provide such a guarantee.
+		// In that case, converting a CStr to str will return a Utf8Error.
+		//
+		// See: https://nrc.github.io/big-book-ffi/reference/strings.html
+		let host_path_option = file_map.get_host_path(guest_path);
+		if let Some(host_path) = host_path_option {
+			// This variable has to exist, as pointers don't have a lifetime
+			// and appending .as_ptr() would lead to the string getting
+			// immediately deallocated after the statement. Nothing is
+			// referencing it as far as the type system is concerned".
+			//
+			// This is also why we can't just have one unsafe block and
+			// one path variable, otherwise we'll get a use after free.
+			let host_path_c_string = CString::new(host_path.as_bytes()).unwrap();
+			let new_host_path = host_path_c_string.as_c_str().as_ptr();
+
+			unsafe {
+				sysopen.ret = libc::open(new_host_path, flags, sysopen.mode);
+			}
```

Comment:
I think this can be more concise:
```suggestion
			let host_path_c_string = CString::new(host_path.as_bytes()).unwrap();
			// Safety: host_path_c_string is a valid c string.
			unsafe {
				sysopen.ret = libc::open(host_path_c_string.as_c_str().as_ptr(), flags, sysopen.mode);
			}
```